src/Workspaces/Core/Portable/Workspace/WorkspaceEventMap.cs

```diff
@@ -0,0 +1,129 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis;
+using Roslyn.Utilities;
+using static Microsoft.CodeAnalysis.Workspace;
+
+namespace Microsoft.CodeAnalysis;
+
+internal sealed class WorkspaceEventMap
+{
+    private readonly SemaphoreSlim _guard = new(initialCount: 1);
+    private readonly Dictionary<WorkspaceEventType, EventHandlerSet> _eventTypeToHandlerSet = [];
+
+    public WorkspaceEventRegistration AddEventHandler(WorkspaceEventType eventType, WorkspaceEventHandlerAndOptions handlerAndOptions)
+    {
+        using (_guard.DisposableWait())
+        {
+            _eventTypeToHandlerSet[eventType] = _eventTypeToHandlerSet.TryGetValue(eventType, out var handlers)
+                ? handlers.AddHandler(handlerAndOptions)
+                : EventHandlerSet.Empty.AddHandler(handlerAndOptions);
+        }
+
+        return new WorkspaceEventRegistration(this, eventType, handlerAndOptions);
+    }
+
+    public void RemoveEventHandler(WorkspaceEventType eventType, WorkspaceEventHandlerAndOptions handlerAndOptions)
+    {
+        using (_guard.DisposableWait())
+        {
+            _eventTypeToHandlerSet.TryGetValue(eventType, out var originalHandlers);
+
+            // An earlier AddEventHandler call would have created the WorkspaceEventRegistration whose
+            // disposal would have called this method.
+            Debug.Assert(originalHandlers != null);
+
+            if (originalHandlers != null)
+            {
+                var newHandlers = originalHandlers.RemoveHandler(handlerAndOptions);
+                Debug.Assert(originalHandlers != newHandlers);
+
+                _eventTypeToHandlerSet[eventType] = newHandlers;
+            }
```

Comment:
this coul dbe `GetEventHandlerSet(eventType).RemoveHandler(...)` and just assert that was different.